app/invocation/invocation_action_card.tsx

```diff
@@ -587,6 +587,126 @@ export default class InvocationActionCardComponent extends React.Component<Props
     );
   }
 
+  private renderExpectedOutputs(command: build.bazel.remote.execution.v2.Command) {
+    const useOutputPaths =
+      command.outputPaths.length && !(command.outputFiles.length + command.outputDirectories.length);
+
+    const renderOutputPaths = () => (
+      <>
+        {command.outputPaths.map((expectedOutput) => (
+          <div className="expected-output">
+            <span>
+              <FileQuestion className="icon file-question-icon" />
+            </span>
+            <span className="expected-output-label">{expectedOutput}</span>
+          </div>
+        ))}
+      </>
+    );
+
+    const renderOutputFilesAndDirs = () => (
+      <>
+        {command.outputDirectories.map((expectedDir) => (
+          <div className="expected-output">
+            <span>
+              <Folder className="icon folder-icon" />
+            </span>
+            <span className="expected-output-label">{expectedDir}</span>
+          </div>
+        ))}
+        {command.outputFiles.map((expectedFile) => (
+          <div className="expected-output">
+            <span>
+              <File className="icon file-icon" />
+            </span>
+            <span className="expected-output-label">{expectedFile}</span>
+          </div>
+        ))}
+      </>
+    );
+
+    return (
+      <div className="action-section">
+        <div className="action-property-title">Expected Outputs</div>
+        <div className="action-list">{useOutputPaths ? renderOutputPaths() : renderOutputFilesAndDirs()}</div>
+      </div>
+    );
+  }
+
+  private renderMissingOutputs(
+    command: build.bazel.remote.execution.v2.Command,
+    actionResult: build.bazel.remote.execution.v2.ActionResult
+  ) {
+    const useOutputPaths =
+      command.outputPaths.length && !(command.outputFiles.length + command.outputDirectories.length);
+
+    const renderMissingOutputPaths = (missingOutputs: Array<string>) => (
+      <>
+        {missingOutputs.map((missingOutput) => (
+          <div className="missing-output">
+            <span>
+              <FileQuestion className="icon file-question-icon" />
+            </span>
+            <span className="missing-output-label">{missingOutput}</span>
+          </div>
+        ))}
+      </>
+    );
+
+    const renderMissingOutputFilesAndDirs = (missingFiles: Array<string>, missingDirs: Array<string>) => (
+      <>
+        {missingDirs.map((missingDir) => (
+          <div className="missing-output">
+            <span>
+              <Folder className="icon file-question-icon" />
+            </span>
+            <span className="missing-output-label">{missingDir}</span>
+          </div>
+        ))}
+        {missingFiles.map((missingFile) => (
+          <div className="missing-output">
+            <span>
+              <FileQuestion className="icon file-question-icon" />
```

Comment:
(per other commment)

```suggestion
              <FileQuestion className="icon file-question-icon icon-red" />
```